### DISCUSSION TITLE
Configurable aspect ratio for the camera image

### DISCUSSION THREAD

**Discussion by FHeilmann** [sca075/mqtt_vacuum_camera#61] [Ideas]
### Description

Hi, first of all, thanks for the absolutely amazing addon that you provided here. It replaced a bunchof manual work I needed to do to enable using my Valetudo Vacuum with the `lovelace-xiaomi-vacuum-map-card`!

I have the card set up on my dashboard together with a few buttons and stats cards to display things like RSSI, consumables, etc. Unfortunately, on one of my wallmounted tablets, this causes the resulting card to exceed the available vertical screen space. I have since used `card-mod` to tweak the aspect ratio of the map image, but I'd prefer to be able to use the configuration of this integration to set an aspect ratio for the camera image (e.g. 2:1) or explicit width and height (e.g. 400x200).

Thank you for considering this feature request.

Florian

### Solution

Integration allows to set aspect ratio of the camera image, resulting image provided to home assistant is padded to achieve desired aspect ratio

### Alternatives

Integration allows to set explicit width and height of the camera image

**Comment by sca075** [comment]
@FHeilmann it is indeed an interesting the idea.. but please let's make a point here too.
- On earlier versions of the camera there was crop and trim to resize the image as desired.
- On the newer versions, as I guess you have seen, we use I kind of auto trimming (that of course also crops the images). 
Reducing the aspect ratio of the image would also improve drastically the performances of the camera.. anyhow as you wrote from the card of Piotr (well I'm in contact with him and actually I did work out some changes on the card as well to grab the topic of the vacuums directly from the camera).. is possible actually to setup as you did the image at a specific aspect ratio.
- If we do this kind of modification.. it will be difficult to calculate the calibration_points. Now I do confess that it is a good proposal. But please understand in this moment will be a little difficult to put it in place.. this do not mean that is impossible. 

As soon I can study an how to.. but please understand right now I do have to address other basic function first, I would come back on it.. anyhow if you are willing to contribute.. you can always fork the project and of course you will found my full cooperation if you would start or have in mind how to design such function... for the time be I would propose we keep this as an open discussion if you agree ;)  when you agree I will move it to the discussions.

**Comment by FHeilmann** [comment]
Thank you for your timely and thorough response. 

I'm fine with however way you want to approach the issue. Given that I "solved the issue" by modifying piotr's card it might make sense to make this change in the card instead. I might reach out to him about that. 

I'm fine with moving this to discussions. Also please don't feel rushed in any way. 

Bests,

Florian

**Comment by sca075** [comment]
> Thank you for your timely and thorough response.
> 
> I'm fine with however way you want to approach the issue. Given that I "solved the issue" by modifying piotr's card it might make sense to make this change in the card instead. I might reach out to him about that.
> 
> I'm fine with moving this to discussions. Also please don't feel rushed in any way.
> 
> Bests,
> 
> Florian

Florian, I will work on the card too, and discuss about it with Piotr (that is at the moment quite busy). Anyhow I will keep this open as per the idea you had is quite interesting to me, this would solve an issue that other users have using the VM's instead of real hardware... mmmmm we will see, but please I will need a little time as above explained for it ;)

**Comment by sca075** [comment]
@FHeilmann 
Dear Florian,
Wishing you all the best and a serine xMas time, I just had an idea that I would like to share with you. At current the camera is detecting the edges of the image as explained above. I'm thinking to make so that as soon the rooms are detected the image trimming will be update with the rooms layout data. This should remove the lidar imperfections and reduce already the images sizes a lot.
We do use to draw the map also pixel_size that could be used to "re-scale" the image, actually changing the pixel size of the rendered image result in a smaller image too (from some test I did).
The standard pixel_sizes are 5 or 7 according to the default values Valetudo that I've collected but of course this do not matter, I mean we receive it with the image data from the Vacuum, and we can use a scale variable that default would be the standard pixel_value to scale factor.
This should allow to implement the function you request on the camera without to compromise the calibration_points, I mean I think make possible to change the image ratio and implement it by the images option. 
I will start to work on this direction for the required modifications.. and open a PR for this as per once I make stable the 1.5.x this can be implemented as a major change (if it will work). 
Once Again thanks a lot for the Valuable advise here, all the best,
Sandro

**Comment by FHeilmann** [comment]
Dear @sca075, thank you and all the best to you and yours as well. I just realized we had a bit of a misunderstanding. 

I'm not talking about scaling the image, im talking about changing the aspect ratio, padding one dimension so that the image becomes wider or taller as a result. Let me show you what I mean with a screenshot from my current setup (modded with card mod) 

![PXL_20231224_122742564 MP](https://github.com/sca075/valetudo_vacuum_camera/assets/4352664/ce0c69af-f618-4e29-b371-6dfddbca12b5)


As you can see the camera image does not occupy the entire width of the card (and therefore the image is less tall and fits on the screen without scrolling. This would've been easier if I could change the aspect ratio of the image (from the rectangular 1:1 to, e.g. 2:1 in my case)

Hope this clarifies my issue. 

All the best.
Florian

**Comment by sca075** [comment]
got it but let me clarify the pixel_size is actually the image aspect ratio.. it means the image will reduce in size too.. anyhow the card is going to make it full size, therefore zooming out two times and locking the aspect ratio will result on the same as per your example, I will work on it the next year ;)

**Reply by sca075** [reply]
@FHeilmann I did work mainly on the auto zoom recently and I guess I found a way to get the aspect ratio in the options of the camera as should it be. I'm really sorry about the long time waiting.. but I didn't forget about this.
If you would be so kind to start this project.. thank once again for the idea.

**Comment by sca075** [comment]
@FHeilmann in the V1.6.0 there will be the possibility to lock the aspect ratio directly in the camera options.. as soon I'm done with it and (right now I'm in test phase of this) the function will be available. If you are still using this integration please consider to star the project it helps really a lot.
Thanks,
Sandro

**Reply by FHeilmann** [reply]
Gladly, thanks so much for working on this.

**Reply by sca075** [reply]
Dear @FHeilmann
This a short update about the tests done.
- It is possible to scale down the image and maintain the calibration data 1:2, 1:3, 1:4 and 1:8.
- if I change the aspect ratio of the image (test done with 1:1, 2:1) then the calibration data need to be re-calculated in a different way (basically I need to change drastically the logic behind this).
- In both of the test done.. Home Assistant by default will force he image to be display the image at full card size.
Indeed I think to release the option simply to scale down the image to a different size.. below examples of the images scaled.
original size:
![snapshot_models5 (2)](https://github.com/sca075/valetudo_vacuum_camera/assets/82227818/29c89686-66fa-4229-8c2d-9a43646f429b)
scaled 1:2:
![snapshot_models5 (3)](https://github.com/sca075/valetudo_vacuum_camera/assets/82227818/a35e7e53-f8f2-486b-ad50-84d130f2604b)
scaled 1:8:
![snapshot_models5 (4)](https://github.com/sca075/valetudo_vacuum_camera/assets/82227818/da832e35-d030-49b5-b2d9-d9b80a4c28e5)

To be honest this the best I think I can do, anyhow I think probably it can help as per your initial proposal was to resize the image to a specific value. The option to lock the aspect ratio will consider the initial image size aspect ratio and keep it constant while resize of the image is done. If aspect ratio lock is disabled with auto-zoom enable the aspect ratio may change in base of the shape and size of the room. 

Will do some additional test by padding the images so that we can probably full fill the spec. this require some additional steps in the image processing then please allow me to reserve to it some additional time.

Thanks in advance for you kind understanding, hopefully this can helps to feet the image in the card (fully display in it) without to crash the layout of the tiles.

Regards,
Sandro

**Reply by sca075** [reply]
it looks I got it working keeping the desired aspect ratio and padding the image :) thanks a lot for the idea

**Reply by sca075** [reply]
and this is the result at 2:1 aspect ratio :)
![Screenshot 2024-03-13 at 15 54 29](https://github.com/sca075/valetudo_vacuum_camera/assets/82227818/d3782e9e-38a6-48b9-bd00-0324adcdc09b)

**Comment by sca075** [comment]
Dear @FHeilmann,

https://github.com/sca075/valetudo_vacuum_camera/discussions/96 a beta release of the v1.6.0 was just released and should be possible to download it via HACS.

![Screenshot 2024-03-15 at 15 28 28](https://github.com/sca075/valetudo_vacuum_camera/assets/82227818/cb84530b-dc45-427b-bd6b-d10cb3e77ac4)

Would be really nice of you when you can confirm that the aspect ratio 2:1 as shown above fit your dashboard.
Thanks in advance, and wishing you a nice weekend.
Sandro

**Comment by FHeilmann** [comment]
Hey @sca075 . Unfortunately my vacuum gave up the ghost recently so I currently don't have a valetudo platform to test this with. I will definitely report back once I'm back up and running again!
